### PR TITLE
(RHEL-16182) Support remote sealing in systemd-cryptenroll/systemd-cryptsetup

### PR DESCRIPTION
**Pull request by dtardon**
Resolves: RHEL-16182

### CODE REVIEW

**Review by mrc0mmand — CHANGES_REQUESTED** [review]

File: test/TEST-70-TPM2/test.sh [hunk 1]
```diff
@@ -12,6 +12,12 @@ TEST_REQUIRE_INSTALL_TESTS=0
 
 command -v swtpm >/dev/null 2>&1 || exit 0
 command -v tpm2_pcrextend >/dev/null 2>&1 || exit 0
+command -w openssl >/dev/null 2>&1 || exit 0
```
[the remark this answers is not in the document]
Comment (dtardon) [reply test/TEST-70-TPM2/test.sh]: I wish I could say I made the typo intentionally to get the test silently skipped. I've no idea how did it happen...

**Comment by mrc0mmand** [comment]
@dtardon could you do one (hopefully) last force-push, just to trigger all CIs again, since the last results are ~month old?

**Comment by mrc0mmand** [comment]
Hmpf, I was just about to press the approve button, but I noticed one potential issue - I was running _everything_ with sanitizers "locally" (on a hefty Beaker machine) and, for some reason, TEST-70-TPM2 fails there - but _only_ under sanitizers:

```
# journalctl -b -o short-monotonic --no-hostname --file /var/tmp/systemd-test.vjb6xq/system.journal -p info _SYSTEMD_UNIT=testsuite-70.service 
...
[    4.666325] testsuite-70.sh[575]: + img=/var/tmp/test.img
[    4.666325] testsuite-70.sh[575]: + truncate -s 20M /var/tmp/test.img
[    4.681569] testsuite-70.sh[575]: + echo -n passphrase
[    4.681569] testsuite-70.sh[575]: + cryptsetup luksFormat -q --pbkdf pbkdf2 --pbkdf-force-iterations 1000 --use-urandom /var/tmp/test.img /tmp/passphrase
[    5.123436] testsuite-70.sh[575]: + systemd-cryptenroll --unlock-key-file=/tmp/passphrase --tpm2-device=auto /var/tmp/test.img
[    5.229093] testsuite-70.sh[580]: Allocating context for crypt device /var/tmp/test.img.
[    5.233294] testsuite-70.sh[580]: Trying to open and read device /var/tmp/test.img with direct-io.
[    5.237233] testsuite-70.sh[580]: Initialising device-mapper backend library.
[    5.238106] testsuite-70.sh[580]: Trying to load LUKS2 crypt type from device /var/tmp/test.img.
[    5.241297] testsuite-70.sh[580]: Crypto backend (OpenSSL 3.0.7 1 Nov 2022 [default][legacy]) initialized in cryptsetup library version 2.6.0.
[    5.244724] testsuite-70.sh[580]: Detected kernel Linux 5.14.0-415.el9.x86_64 x86_64.
[    5.246809] testsuite-70.sh[580]: Loading LUKS2 header (repair disabled).
[    5.247598] testsuite-70.sh[580]: Acquiring read lock for device /var/tmp/test.img.
[    5.248476] testsuite-70.sh[580]: Verifying lock handle for /var/tmp/test.img.
[    5.249771] testsuite-70.sh[580]: Device /var/tmp/test.img READ lock taken.
[    5.249771] testsuite-70.sh[580]: Trying to read primary LUKS2 header at offset 0x0.
[    5.249771] testsuite-70.sh[580]: Opening locked device /var/tmp/test.img
[    5.249771] testsuite-70.sh[580]: Verifying locked device handle (regular file)
[    5.254125] testsuite-70.sh[580]: LUKS2 header version 2 of size 16384 bytes, checksum sha256.
[    5.260175] testsuite-70.sh[580]: Checksum:7317a11736e33a104d9dfb549e3ad8af94158e128372cc86010652b47d4fbd04 (on-disk)
[    5.260175] testsuite-70.sh[580]: Checksum:7317a11736e33a104d9dfb549e3ad8af94158e128372cc86010652b47d4fbd04 (in-memory)
[    5.260175] testsuite-70.sh[580]: Trying to read secondary LUKS2 header at offset 0x4000.
[    5.260175] testsuite-70.sh[580]: Reusing open ro fd on device /var/tmp/test.img
[    5.269115] testsuite-70.sh[580]: LUKS2 header version 2 of size 16384 bytes, checksum sha256.
[    5.271208] testsuite-70.sh[580]: Checksum:f133cb3ba19e54a8c3154d9c0782bb8f7b75b9c777ba2061c1703212b95407a2 (on-disk)
[    5.271208] testsuite-70.sh[580]: Checksum:f133cb3ba19e54a8c3154d9c0782bb8f7b75b9c777ba2061c1703212b95407a2 (in-memory)
[    5.271208] testsuite-70.sh[580]: Device size 20971520, offset 16777216.
[    5.271208] testsuite-70.sh[580]: Device /var/tmp/test.img READ lock released.
[    5.271208] testsuite-70.sh[580]: Only 1 active CPUs detected, PBKDF threads decreased from 4 to 1.
[    5.271208] testsuite-70.sh[580]: PBKDF argon2id, time_ms 2000 (iterations 0), max_memory_kb 1048576, parallel_threads 1.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 0.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 1.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 2.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 3.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 4.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 5.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 6.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 7.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 8.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 9.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 10.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 11.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 12.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 13.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 14.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 15.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 16.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 17.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 18.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 19.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 20.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 21.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 22.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 23.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 24.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 25.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 26.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 27.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 28.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 29.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 30.
[    5.271208] testsuite-70.sh[580]: Requesting JSON for token 31.
[    5.271208] testsuite-70.sh[580]: /tmp/passphrase has 0644 mode that is too permissive, please adjust the ownership and access mode.
[    5.271208] testsuite-70.sh[580]: Keyslot 0 priority 1 != 2 (required), skipped.
[    5.271208] testsuite-70.sh[580]: Trying to open LUKS2 keyslot 0.
[    5.271208] testsuite-70.sh[580]: Running keyslot key derivation.
[    5.337915] testsuite-70.sh[580]: Reading keyslot area [0x8000].
[    5.337915] testsuite-70.sh[580]: Acquiring read lock for device /var/tmp/test.img.
[    5.337915] testsuite-70.sh[580]: Verifying lock handle for /var/tmp/test.img.
[    5.337915] testsuite-70.sh[580]: Device /var/tmp/test.img READ lock taken.
[    5.337915] testsuite-70.sh[580]: Reusing open ro fd on device /var/tmp/test.img
[    5.337915] testsuite-70.sh[580]: Device /var/tmp/test.img READ lock released.
[    5.416650] testsuite-70.sh[580]: Verifying key from keyslot 0, digest 0.
[    5.420784] testsuite-70.sh[580]: Digest 0 (pbkdf2) verify failed with -1.
[    5.424231] testsuite-70.sh[580]: Unlocking via keyfile failed: Operation not permitted
[    5.425169] testsuite-70.sh[580]: Releasing crypt device /var/tmp/test.img context.
[    5.426063] testsuite-70.sh[580]: Releasing device-mapper backend.
[    5.426876] testsuite-70.sh[580]: Closing read only fd for /var/tmp/test.img.
```

However, running the same test with upstream systemd (also with sanitizers) works fine. I very vaguely remember dealing with something similar in upstream (https://github.com/systemd/systemd/issues/27716), but in that case the issue was intermittent, here it always reproduces. I'll try to read through _the archives_ to see what's going on.

**Review by mrc0mmand — APPROVED** [review]
LGTM, but setting the "dont-merge" flag so it won't get merged until @msekletar blesses it as well.

**Comment by mrc0mmand** [comment]
Heh, as for the issue I mentioned in https://github.com/redhat-plumbers/systemd-rhel9/pull/214#issuecomment-1941919896 - I did a quick bisect on the upstream repo and the commit that "fixes" it is 920eb35db99e0e96b5b6b9ba4da37d0b81333276, more specifically it's the part that moves the test image from `/var/tmp/` to `/tmp/`.

With:

```diff
diff --git a/test/units/testsuite-70.sh b/test/units/testsuite-70.sh
index 222fa33d69..b61cb0e572 100755
--- a/test/units/testsuite-70.sh
+++ b/test/units/testsuite-70.sh
@@ -15,7 +15,7 @@ cleanup() {
 persistent=""
 
 # Prepare fresh disk image
-img="/var/tmp/test.img"
+img="/tmp/test.img"
 truncate -s 20M $img
 
 echo -n passphrase >/tmp/passphrase
```

the test passes even with the current branch on my Beaker machine. No idea why it works in CI, but doesn't for me locally, but whatever. (I also like the `No functional changes` part of the commit message ;))

**Review by msekletar — APPROVED** [review]
LGTM

**Comment by jamacku** [comment]
@dtardon there are two follow-ups in upstream. Not sure if they are important:

#### Follow-ups

| commit|follow-up |
|---|---|
| https://github.com/redhat-plumbers/systemd-rhel9/commit/7ac6147b69cd59c636c9ac290a5d0053c54f884d - _generators: optionally, measure file systems at boot_ | https://github.com/systemd/systemd/commit/0eb635ef4bc11792cd4ef384ae252a2c7fd4122a |
| https://github.com/redhat-plumbers/systemd-rhel9/commit/cb40b03557820ed08d46ce172a79ae5af9b0c260 - _tpm2: add support for a trusted SRK_ | https://github.com/systemd/systemd/commit/42559808870e21d1f401f74f0381b88d201cbb3b |

#### Commit mentions

| commit|mention |
|---|---|
| https://github.com/redhat-plumbers/systemd-rhel9/commit/9a13aa0d9d7b73df51da4c1cd2368a2699ec014e - _boot: Simplify object erasure_ | https://github.com/systemd/systemd/commit/5deb391c6e6d2b8fd7b94234efea49cd6bee0d76 |
| https://github.com/redhat-plumbers/systemd-rhel9/commit/173953711b30b968d04b8e9ba6f4bb101a612a47 - _tpm2: add salt to pin_ | https://github.com/systemd/systemd/commit/3c6439bf2ce1afaf6e922f3e0064c14f0c7bfa11 |
| https://github.com/redhat-plumbers/systemd-rhel9/commit/d0e16e7af15dda325fa03483b22cc33d9c1e2d55 - _boot/measure: replace TPM PolicyPCR session with calculation_ | https://github.com/systemd/systemd/commit/0d66468243d888dd721ba0072cbad742ab6fc690 |
| https://github.com/redhat-plumbers/systemd-rhel9/commit/79f5694f8ac8963981134b788d969d080df73279 - _tpm2: add functions to convert TPM2B_PUBLIC to/from openssl pkey or PE…_ | https://github.com/systemd/systemd/commit/fda3e844657b61b69f8d0badffb6239840ef9e97</br>https://github.com/systemd/systemd/commit/1242b9ab2bd306df0df51ca9ee7801f572ce1e28 |
| https://github.com/redhat-plumbers/systemd-rhel9/commit/96dd5b187bb955637d1b60e7922ff969498b3ece - _cryptenroll: add support for calculated TPM2 enrollment_ | https://github.com/systemd/systemd/commit/78fdf0f68b07b0539305b128a00d0f8428a055cd |
| https://github.com/redhat-plumbers/systemd-rhel9/commit/220c9dc7ce3b8008a429e30415fd425d6a9a6f34 - _creds-util: do not try TPM2 if there is not support_ | https://github.com/systemd/systemd/commit/a9e4057f1f49ab3d10a32d898b904c4ccfeeb41e |

**Comment by dtardon** [comment]
> @dtardon there are two follow-ups in upstream. Not sure if they are important:
> #### Follow-ups
> commit 	follow-up
> [7ac6147](https://github.com/redhat-plumbers/systemd-rhel9/commit/7ac6147b69cd59c636c9ac290a5d0053c54f884d) - _generators: optionally, measure file systems at boot_ 	[systemd/systemd@0eb635e](https://github.com/systemd/systemd/commit/0eb635ef4bc11792cd4ef384ae252a2c7fd4122a)

I'm pretty sure I saw this one and skipped it as not needed for RHEL-9's build configuration. CI seems to work too.

> [cb40b03](https://github.com/redhat-plumbers/systemd-rhel9/commit/cb40b03557820ed08d46ce172a79ae5af9b0c260) - _tpm2: add support for a trusted SRK_ 	[systemd/systemd@4255980](https://github.com/systemd/systemd/commit/42559808870e21d1f401f74f0381b88d201cbb3b)

This, again, is not needed for our build config.

**Pull request merged**